### PR TITLE
feat: make multiple controllers handle different ApisixRoute CRDs

### PR DESCRIPTION
**Pull request by Donghui0**
Please answer these questions before submitting a pull request

- Why submit this pull request?
- [ ] Bugfix
- [x] New feature provided
- [ ] Improve performance
- [ ] Backport patches

- Related issues
#578 
#592 
___

___
### New feature or improvement
make multiple controllers handle different ApisixRoute CRDs

___
### Backport patches
- Why need to backport?

- Source branch

- Related commits and pull requests

- Target branch

### CODE REVIEW

File: cmd/ingress/ingress.go [hunk 1]
```diff
@@ -144,6 +144,7 @@ the apisix cluster and others are created`,
 	cmd.PersistentFlags().DurationVar(&cfg.Kubernetes.ResyncInterval.Duration, "resync-interval", time.Minute, "the controller resync (with Kubernetes) interval, the minimum resync interval is 30s")
 	cmd.PersistentFlags().StringSliceVar(&cfg.Kubernetes.AppNamespaces, "app-namespace", []string{config.NamespaceAll}, "namespaces that controller will watch for resources")
 	cmd.PersistentFlags().StringVar(&cfg.Kubernetes.IngressClass, "ingress-class", config.IngressClass, "the class of an Ingress object is set using the field IngressClassName in Kubernetes clusters version v1.18.0 or higher or the annotation \"kubernetes.io/ingress.class\" (deprecated)")
+	cmd.PersistentFlags().StringVar(&cfg.Kubernetes.ApisixRouteClass, "apisix-route-class", config.ApisixRouteClass, "the class of an ApisixRoute object is set using the field ApisixRouteClassName in Kubernetes")
```
Comment (tokers) [inline cmd/ingress/ingress.go:147]: We may just use `IngressClass` field to manage all resources.

**Comment by gxthrj** [comment]
Here is an [issue](https://github.com/apache/apisix-ingress-controller/issues/592) related

File: cmd/ingress/ingress.go (hunk 1 above)
Comment (Donghui0) [reply cmd/ingress/ingress.go:147]: Ok. I will fix it later.

File: cmd/ingress/ingress.go [hunk 2]
```diff
@@ -143,8 +143,7 @@ the apisix cluster and others are created`,
 	cmd.PersistentFlags().StringVar(&cfg.Kubernetes.Kubeconfig, "kubeconfig", "", "Kubernetes configuration file (by default in-cluster configuration will be used)")
 	cmd.PersistentFlags().DurationVar(&cfg.Kubernetes.ResyncInterval.Duration, "resync-interval", time.Minute, "the controller resync (with Kubernetes) interval, the minimum resync interval is 30s")
 	cmd.PersistentFlags().StringSliceVar(&cfg.Kubernetes.AppNamespaces, "app-namespace", []string{config.NamespaceAll}, "namespaces that controller will watch for resources")
-	cmd.PersistentFlags().StringVar(&cfg.Kubernetes.IngressClass, "ingress-class", config.IngressClass, "the class of an Ingress object is set using the field IngressClassName in Kubernetes clusters version v1.18.0 or higher or the annotation \"kubernetes.io/ingress.class\" (deprecated)")
-	cmd.PersistentFlags().StringVar(&cfg.Kubernetes.ApisixRouteClass, "apisix-route-class", config.ApisixRouteClass, "the class of an ApisixRoute object is set using the field ApisixRouteClassName in Kubernetes")
```
Comment (tokers) [inline cmd/ingress/ingress.go]: This is a big broken change, we need some discussions in the mailing list.

File: pkg/ingress/apisix_route.go [hunk 3]
```diff
@@ -292,6 +292,26 @@ func (c *apisixRouteController) handleSyncErr(obj interface{}, errOrigin error)
...
+	var class string
+
+	if ar.GroupVersion() == kube.ApisixRouteV1 {
+		class = ar.V1().Spec.IngressClassName
+	} else if ar.GroupVersion() == kube.ApisixRouteV2alpha1 {
+		class = ar.V2alpha1().Spec.IngressClassName
+	} else if ar.GroupVersion() == kube.ApisixRouteV2beta1 {
+		class = ar.V2beta1().Spec.IngressClassName
+	} else {
+		class = c.controller.cfg.Kubernetes.IngressClass
```
Comment (tokers) [inline pkg/ingress/apisix_route.go:305]: Actually the else won't be entered, once the logic here was executed, then there are some programming faults occurred, we may just:

1. Panic, or
2. log the error and just return false.

**Review by tokers — APPROVED** [review]

**Comment by tao12345666333** [comment]
Thanks for your contribution,
Please resolve the conflict if you have time,
let's move forward.

**Comment by tao12345666333** [comment]
there are some conflicts, and I want the feature to just be added to the v2 version of the API

https://github.com/apache/apisix-ingress-controller/issues/592#issuecomment-1221021549

**Comment by Gallardot** [comment]
@Donghui0  Do you have time to work on this PR? Maybe I can help with that.

**Comment by tao12345666333** [comment]
Yesterday I had an online meeting with @Donghui0  and @Gallardot.
After discussion, we will use this PR instead of #1523, and @Donghui0 will complete this PR. 

I've just resolved the code merge conflict in this PR.

For some basic technical solutions can refer to  https://github.com/apache/apisix-ingress-controller/pull/1523#issuecomment-1429344801

I will also submit a more complete description at https://github.com/apache/apisix-ingress-controller/issues/592 later

Thank you all!

**Comment by tao12345666333** [comment]
@Donghui0 Please see https://github.com/apache/apisix-ingress-controller/issues/592#issuecomment-1449264770 
I have added more details.

File: pkg/kube/apisix/apis/config/v2beta3/types.go [hunk 4]
```diff
@@ -44,8 +44,9 @@ type ApisixStatus struct {
 
 // ApisixRouteSpec is the spec definition for ApisixRouteSpec.
 type ApisixRouteSpec struct {
-	HTTP   []ApisixRouteHTTP   `json:"http,omitempty" yaml:"http,omitempty"`
-	Stream []ApisixRouteStream `json:"stream,omitempty" yaml:"stream,omitempty"`
+	IngressClassName string              `json:"ingressClassName,omitempty" yaml:"ingressClassName,omitempty"`
+	HTTP             []ApisixRouteHTTP   `json:"http,omitempty" yaml:"http,omitempty"`
+	Stream           []ApisixRouteStream `json:"stream,omitempty" yaml:"stream,omitempty"`
```
Comment (tao12345666333) [inline pkg/kube/apisix/apis/config/v2beta3/types.go:49]: We don't need to change the v2beta3 resource.

File: pkg/providers/apisix/apisix_route.go [hunk 5]
```diff
@@ -534,6 +534,23 @@ func (c *apisixRouteController) handleSyncErr(obj interface{}, errOrigin error)
 	c.MetricsCollector.IncrSyncOperation("route", "failure")
 }
 
+func (c *apisixRouteController) isEffective(ar kube.ApisixRoute) bool {
+	var icn string
+
+	switch ar.GroupVersion() {
+	case config.ApisixV2beta3:
+		icn = ar.V2beta3().Spec.IngressClassName
```
Comment (tao12345666333) [inline pkg/providers/apisix/apisix_route.go:542]: ditto

**Review by tao12345666333 — COMMENTED** [review]
And please add e2e test cases to cover this change. Thanks

File: samples/deploy/crd/v1/ApisixRoute.yaml [hunk 6]
```diff
@@ -69,6 +69,8 @@ spec:
                 - required: ["http"]
                 - required: ["stream"]
               properties:
+                ingressClassName:
+                  type: string
```
Comment (lingsamuel) [inline samples/deploy/crd/v1/ApisixRoute.yaml:73]: no longer need this

**Review by lingsamuel — CHANGES_REQUESTED** [review]

**Review by lingsamuel — APPROVED** [review]

File: go.mod [hunk 7]
```diff
@@ -28,6 +28,11 @@ require (
 	sigs.k8s.io/yaml v1.3.0
 )
 
+require (
+	github.com/go-task/slim-sprig v0.0.0-20210107165309-348f09dbbbc0 // indirect
+	github.com/google/pprof v0.0.0-20210407192527-94a9f03dee38 // indirect
+)
+
```
Comment (tao12345666333) [inline go.mod:35]: you don't need to change this one. right?

**Review by tao12345666333 — COMMENTED** [review]
Please resolve the conflicts, thanks

**Comment by tao12345666333** [comment]
CI fails, I think @AlinsRan  can help. PTAL

**Review by Gallardot — APPROVED** [review]
LGTM

File: go.mod [hunk 8]
```diff
@@ -67,6 +67,8 @@ require (
 	github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd // indirect
 	github.com/modern-go/reflect2 v1.0.2 // indirect
 	github.com/munnerz/goautoneg v0.0.0-20191010083416-a7dc8b61c822 // indirect
+	github.com/onsi/ginkgo/v2 v2.9.0 // indirect
+	github.com/onsi/gomega v1.27.1 // indirect
```
Comment (AlinsRan) [inline go.mod:71]: These two dependencies are redundant and will only be used in e2e. Let me remove them.

**Pull request merged**